### ISSUE TITLE
Feature Request: Module versioning for git sources

### ISSUE THREAD

**Issue by apottere** [opentofu/opentofu#1017]
### OpenTofu Version

```shell
OpenTofu v1.6.0-beta4
on darwin_arm64
```

### Use Cases

As someone who uses git-based modules extensively for distributing proprietary terraform modules, I often find myself in a situation where I have many modules that all reference the same git repo (with subdirectories) at a certain `ref`.  When I publish a new tag of the module, it's frustrating to have to find/replace every instance of the old tag with the new tag.  I would like a single source of truth for a ref tag for a certain git module.

### Attempted Solutions

This is not possible without find/replace or a script that does find/replace for you as far as I'm aware.

### Proposal

The crux of the problem seems to be that module sources can't use traditional variables, since those variables are dependent on knowing which module to download.  I don't see a simple way to resolve that, which is why this issue hasn't been fixed yet.

Note that the below proposals apply to a module's `source` string, but could also apply to a module's `version` parameter _if that was updated to set the `ref` parameter for git modules_

Some potential fixes might include:
1. Create a new variable type that can't use interpolated values (e.g. `const`) that can be interpolated into `source` or used on the rhs of `version`
2. Have some constant map of module url prefix => module version somewhere central, such as the `terraform` block.  This is less ideal for a scenario where different modules in the same repo need to use different versions, because of a breaking change etc.
3. Allow injecting environment variables, since those should never change within the scope of the running process (not ideal, since it wouldn't be checked in anywhere)

Examples:
1.
```hcl
const {
    myorg-module-version = "1.1.0"
}
module "function" {
    source = "git::ssh://git@github.com/myorg/terraform-modules.git//function?ref=${const.myorg-module-version}"
    // or
    version = const.myorg-module-version
}
```

2.
```hcl
terraform {
  module_versions {
    git {
      url = "github.com/myorg/terraform-modules//function"
      ref = "1.0.0"
    }
    git {
      url = "github.com/myorg/terraform-modules"
      ref = "1.1.0"
    }
  }
}
module "function" {
    source = "git::ssh://git@github.com/myorg/terraform-modules.git//function" // ref=1.0.0
}
module "bucket" {
    source = "git::ssh://git@github.com/myorg/terraform-modules.git//bucket" // ref=1.1.0
}
```

3.
```hcl
module "function" {
    source = "git::ssh://git@github.com/myorg/terraform-modules.git//function?ref=${env.TF_MYORG_MODULE_VERSION}"
}
```

### References

This has been a request in the terraform community since 2015, and has not been solved.

- https://github.com/hashicorp/terraform/issues/17994
- https://github.com/hashicorp/terraform/issues/1439
- https://github.com/hashicorp/terraform/issues/15613

**Comment by apottere** [comment]
It seems like the only backwards-compatible proposal would be #2 - it wouldn't actively _break_ running terraform but it _would_ potentially produce different results.

**Comment by cube2222** [comment]
Hey, thanks for submitting the issue @apottere!

I definitely think this is a topic worth discussing. Funnily, we've discussed this very topic recently in the core team, and similar ideas have been raised. That said, I'm honestly not a fan of any of the solutions proposed.

The const approach is a huge cannon to throw at this, and would be a big language change, that would also make code incompatible with legacy Terraform. We've agreed to avoid those, at least initially, but in this case I don't think it'd be worth it either way. It would introduce an additional evaluation step during module scanning (a lot of complexity!) and it's non-obvious to me how those would be passed to submodules which might also reference modules themselves.

The env var approach seems hacky to me and would work in a very weird way when used with external modules (external modules referencing arbitrary environment variables on your system for version constraints).

The module_versions approach could cause unintended consequences, as different modules could intentionally be referencing different versions of (other) modules. I could totally see a situation where you want to make sure module A and B are synchronized via that file and reference some source module abc, but there's external module C that uses module D that references the source module abc too, and your change intended to affect A and B will break module D, and thus C.

I don't have an answer here, but language changes like these need to be considered very carefully.

---

One idea I have, is to maybe introduce a concept of "tags" (do not confuse with git tags!) that you could use in version strings of modules. Then, you could use an approach similar to your module_versions approach, but would instead only define what a specific tag means for a given module source. This way you could keep using multiple distinct versions of a single module without issues, and keep the management of the once you want to in a single place. But that's not a fleshed-out proposal, just a loose idea.

**Comment by apottere** [comment]
Thanks for the quick response!

I agree that none of the proposals are perfect, especially w/r/t submodules and backwards-compatibility.

By "tags", do you mean something like:
```hcl
terraform {
    module_version_tags {
        myorg-v1 = "1.1.0"
        myorg-v2 = "2.0.0"
    }
}

module "function" {
    source = "git::ssh://git@github.com/myorg/terraform-modules.git//function?ref=[myorg-v1]" // ref=1.1.0
}
module "bucket" {
    source = "git::ssh://git@github.com/myorg/terraform-modules.git//bucket?ref=[myorg-v2]" // ref=2.0.0
}
```

**Comment by cam72cam** [comment]
I think the tags idea is interesting, kind of a restricted version of the const proposal.  The trick is where the modules are evaluated vs where they are fetched.

Right now the module source is parsed during the hcl -> config step.  Much later on, the config is then used to fetch the modules.  There's room in between the two steps for adding some logic.

In more general terms

RootModule
* Load HCL
* HCL -> Config
* `<INSERT PROPOSAL>`
* Download Referenced Modules
* For each referenced module:
  - Load HCL
  - HCL -> Config
  - `<INSERT PROPOSAL>`
  - Download Referenced Modules
  - Recurse for each referenced modules.

**Comment by cam72cam** [comment]
Also I think this would want to be done without breaking the module -> child module barrier.  Whichever solution is attempted, it should be self-contained to the hcl/config of that module.

**Comment by apottere** [comment]
Agreed - I think it would confusing for any versions to propagate to submodules implicitly.

**Comment by Pikaro** [comment]
If there aren't any regressions, the codebase should already support fetching the download locations of modules via HTTP, and it should support `.netrc`. You can use that to determine which modules to download with a simple API, based on the "username and password" (for us project and environment) in the `.netrc`.

**Comment by cam72cam** [comment]
This is probably related to https://github.com/opentofu/opentofu/issues/450

**Comment by Yantrio** [comment]
This is probably dependant on https://github.com/opentofu/opentofu/issues/1042 also

**Comment by cam72cam** [comment]
Closing as this is now captured in an accepted RFC and will be closed by #286
